presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ReorderWindows.java

```diff
@@ -0,0 +1,152 @@
+ * won't be carried over beyond ProjectNode to precede WindowNode with PartitionBy = A,B:
+ * OutputNode
+ * `--...
+ *    `--WindowNode(PartitionBy A OrderBy F)
+ *       `--ProjectNode(...)
+ *          `--WindowNode(PartitionBy A,B OrderBy C)
+ *             `--...
+ *
+ * The optimizer may reorder windows even if no window partitioning is a prefix of the other.
+ * E.g.:
+ * OutputNode
+ * `--...
+ *    `--WindowNode(PartitionBy A OrderBy B)
+ *       `--WindowNode(PartitionBy C OrderBy D)
+ *          `--...
+ *
+ * May be transformed into:
+ * * OutputNode
+ * `--...
+ *    `--WindowNode(PartitionBy C OrderBy D)
+ *       `--WindowNode(PartitionBy A OrderBy B)
+ *          `--...
+ *
+ * Even though there are no identical `partition by` clauses in such a plan.
+ */
+public class ReorderWindows
+        implements PlanOptimizer
+{
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        return SimplePlanRewriter.rewriteWith(new Rewriter(), plan, new PriorityQueue<>(new PartitionByComparator()));
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<PriorityQueue<WindowNode>>
+    {
+        @Override
+        protected PlanNode visitPlan(PlanNode node, RewriteContext<PriorityQueue<WindowNode>> context)
+        {
+            PriorityQueue<WindowNode> windowNodes = context.get();
+
+            PlanNode sourceNode = context.defaultRewrite(node, new PriorityQueue<>(new PartitionByComparator()));
+            while (windowNodes.peek() != null) {
+                WindowNode windowNode = windowNodes.poll();
+                sourceNode = new WindowNode(
+                        windowNode.getId(),
+                        sourceNode,
+                        windowNode.getSpecification(),
+                        windowNode.getWindowFunctions(),
+                        windowNode.getHashSymbol(),
+                        windowNode.getPrePartitionedInputs(),
+                        windowNode.getPreSortedOrderPrefix());
+            }
+            return sourceNode;
+        }
+
+        @Override
+        public PlanNode visitWindow(WindowNode node, RewriteContext<PriorityQueue<WindowNode>> context)
+        {
+            checkState(!node.getHashSymbol().isPresent(), "ReorderWindows should be run before HashGenerationOptimizer");
+            checkState(node.getPrePartitionedInputs().isEmpty() && node.getPreSortedOrderPrefix() == 0, "ReorderWindows should be run before AddExchanges");
+
+            context.get().add(node);
+            return context.rewrite(node.getSource(), context.get());
+        }
+    }
+
+    private static class PartitionByComparator
+            implements Comparator<WindowNode>
+    {
+        @Override
+        public int compare(WindowNode o1, WindowNode o2)
+        {
+            return stringifyPartitionBy(o1).compareTo(stringifyPartitionBy(o2));
```

Comment:
Instead of converting to strings, compare the lists structurally:

```java
Iterator<Symbol> iterator1 = o1.getPartitionBy().iterator();
Iterator<Symbol> iterator2 = o2.getPartitionBy().iterator();

while (iterator1.hasNext() && iterator2.hasNext()) {
    Symbol symbol1 = iterator1.next();
    Symbol symbol2 = iterator2.next();

    int comparison = symbol1.compareTo(symbol2);
    if (comparison != 0) {
        return comparison;
    }
}

if (iterator1.hasNext()) {
    return 1;
}

if (iterator2.hasNext()) {
    return -1;
}

return 0;
```